File: backend/app/auth/permissions.py
from enum import StrEnum


class Role(StrEnum):
    ADMINISTRATOR = "administrator"
    SECURITY_ANALYST = "security_analyst"
    AUDITOR = "auditor"
    VIEWER = "viewer"


class Permission(StrEnum):
    MANAGE_USERS = "manage_users"
    CONFIGURE_ORGANIZATION = "configure_organization"
    VIEW_SCANS = "view_scans"
    RUN_SCANS = "run_scans"
    ANALYZE_RISKS = "analyze_risks"
    CREATE_MIGRATION_PLANS = "create_migration_plans"
    VIEW_REPORTS = "view_reports"
    EXPORT_FINDINGS = "export_findings"
    VIEW_DASHBOARD = "view_dashboard"


ROLE_PERMISSIONS = {
    Role.ADMINISTRATOR: set(Permission),
    Role.SECURITY_ANALYST: {
        Permission.VIEW_SCANS,
        Permission.RUN_SCANS,
        Permission.ANALYZE_RISKS,
        Permission.CREATE_MIGRATION_PLANS,
        Permission.VIEW_REPORTS,
        Permission.EXPORT_FINDINGS,
        Permission.VIEW_DASHBOARD,
    },
    Role.AUDITOR: {
        Permission.VIEW_SCANS,
        Permission.VIEW_REPORTS,
        Permission.EXPORT_FINDINGS,
        Permission.VIEW_DASHBOARD,
    },
    Role.VIEWER: {Permission.VIEW_SCANS, Permission.VIEW_DASHBOARD},
}


def has_permissions(role: str, permissions: set[Permission]) -> bool:
    try:
        granted = ROLE_PERMISSIONS[Role(role)]
    except (ValueError, KeyError):
        return False
    return permissions.issubset(granted)

File: frontend/src/auth/permissions.ts
import type { UserRole } from "../types/api";

export type Permission =
  | "manage_users"
  | "configure_organization"
  | "view_scans"
  | "run_scans"
  | "analyze_risks"
  | "create_migration_plans"
  | "view_reports"
  | "export_findings"
  | "view_dashboard";

const ALL_PERMISSIONS: Permission[] = [
  "manage_users",
  "configure_organization",
  "view_scans",
  "run_scans",
  "analyze_risks",
  "create_migration_plans",
  "view_reports",
  "export_findings",
  "view_dashboard",
];

/** Mirrors backend/app/auth/permissions.py:ROLE_PERMISSIONS. UI-only — the
 * backend is the real enforcement point; this just decides what to show. */
const ROLE_PERMISSIONS: Record<UserRole, Permission[]> = {
  administrator: ALL_PERMISSIONS,
  security_analyst: [
    "view_scans",
    "run_scans",
    "analyze_risks",
    "create_migration_plans",
    "view_reports",
    "export_findings",
    "view_dashboard",
  ],
  auditor: ["view_scans", "view_reports", "export_findings", "view_dashboard"],
  viewer: ["view_scans", "view_dashboard"],
};

export function hasPermission(role: UserRole, permission: Permission): boolean {
  return ROLE_PERMISSIONS[role]?.includes(permission) ?? false;
}

export function hasAnyRole(role: UserRole, allowed: UserRole[]): boolean {
  return allowed.includes(role);
}
